contracts/exchangeIssuance/FlashMintDex.sol

```diff
@@ -0,0 +1,612 @@
+        uint256 wethReceived = _swapPaymentTokenForWeth(_paymentInfo.token, _paymentInfo.limitAmt, _paymentInfo.swapDataTokenToWeth);
+
+        uint256 wethSpent = _issueExactSetFromWeth(_issueParams);
+        require(wethSpent <= wethReceived, "FlashMint: OVERSPENT WETH");
+        uint256 leftoverWeth = wethReceived.sub(wethSpent);
+        uint256 paymentTokenReturned = 0;
+
+        if (leftoverWeth > 0) {
+            paymentTokenReturned = _swapWethForPaymentToken(leftoverWeth, _paymentInfo.token, _paymentInfo.swapDataWethToToken);
+            _paymentInfo.token.safeTransfer(msg.sender, paymentTokenReturned);
+        }
+
+        paymentTokenSpent = _paymentInfo.limitAmt.sub(paymentTokenReturned);
+
+        emit FlashMint(msg.sender, _issueParams.setToken, _paymentInfo.token, paymentTokenSpent, _issueParams.amountSetToken);
+    }
+
+    /**
+     * Redeems an exact amount of SetTokens for ETH.
+     * The SetToken must be approved by the sender to this contract.
+     *
+     * @param _redeemParams   Struct containing addresses, amounts, and swap data for issuance
+     *
+     * @return ethReceived      Amount of ETH received
+     */
+    function redeemExactSetForETH(IssueRedeemParams memory _redeemParams, uint256 _minEthReceive)
+        external
+        isValidModuleAndSet(_redeemParams.issuanceModule, address(_redeemParams.setToken))
+        nonReentrant
+        returns (uint256 ethReceived)
+    {
+        _redeem(_redeemParams.setToken, _redeemParams.amountSetToken, _redeemParams.issuanceModule);
+
+        ethReceived = _sellComponentsForWeth(_redeemParams);
+        require(ethReceived >= _minEthReceive, "FlashMint: INSUFFICIENT WETH RECEIVED");
+
+        IWETH(WETH).withdraw(ethReceived);
+        payable(msg.sender).sendValue(ethReceived);
+
+        emit FlashRedeem(msg.sender, _redeemParams.setToken, IERC20(ETH_ADDRESS), _redeemParams.amountSetToken, ethReceived);
+        return ethReceived;
+    }
+
+    /**
+     * Redeems an exact amount of SetTokens for an ERC20 token.
+     * The SetToken must be approved by the sender to this contract.
+     *
+     * @param _redeemParams             Struct containing token addresses, amounts, and swap data for issuance
+     *
+     * @return outputTokenReceived      Amount of output token received
+     */
+    function redeemExactSetForERC20(IssueRedeemParams memory _redeemParams, PaymentInfo memory _paymentInfo)
+        external
+        isValidModuleAndSet(_redeemParams.issuanceModule, address(_redeemParams.setToken))
+        nonReentrant
+        returns (uint256 outputTokenReceived)
+    {
+        _redeem(_redeemParams.setToken, _redeemParams.amountSetToken, _redeemParams.issuanceModule);
+
+        uint256 wethReceived = _sellComponentsForWeth(_redeemParams);
+        outputTokenReceived = _swapWethForPaymentToken(wethReceived, _paymentInfo.token, _paymentInfo.swapDataWethToToken);
```

Comment:
Yeah I went with simplicity here rather than trying to optimize for the case that gas costs more than the value of token returned. 

I'll see if I can split out that unused var (I was hitting stack too deep at one point).